java/src/main/java/org/eclipse/ditto/client/messaging/internal/Retry.java

```diff
@@ -0,0 +1,202 @@
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.ditto.client.messaging.internal;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Provides an easy way to implement a an action that should be retried.
+ * In this case this action must return a result which also can be null.
+ * As soon as a result is returned is by the supplier, the action is considered to be finished successfully.
+ *
+ * @param <T> Type of the result
+ */
+final class Retry<T> implements Supplier<CompletionStage<T>> {
+
+    private static final int[] TIME_TO_WAIT_BETWEEN_RETRIES_IN_SECONDS = new int[]{1, 1, 2, 3, 5, 8, 13};
+    private static final Logger LOGGER = LoggerFactory.getLogger(Retry.class.getName());
+
+    private final String sessionId;
+    private final String nameOfAction;
+    private final Supplier<T> retriedSupplier;
+    private final ExecutorService executorService;
+
+
+    private Retry(final String nameOfAction,
+            final String sessionId,
+            final Supplier<T> retriedSupplier,
+            final ExecutorService executorService) {
+
+        this.sessionId = sessionId;
+        this.nameOfAction = nameOfAction;
+        this.retriedSupplier = retriedSupplier;
+        this.executorService = executorService;
+    }
+
+    private static int ensureIndexIntoTimeToWaitBounds(final int index) {
+        if (index < 0) {
+            return 0;
+        } else if (index >= TIME_TO_WAIT_BETWEEN_RETRIES_IN_SECONDS.length) {
+            return TIME_TO_WAIT_BETWEEN_RETRIES_IN_SECONDS.length - 1;
+        } else {
+            return index;
+        }
+    }
+
+    /**
+     * Executes the provided supplier until a result is returned.
+     *
+     * @return A completion stage which finally completes with the result of the supplier. Result can be null.
+     */
+    @Override
+    public CompletionStage<T> get() {
+        return CompletableFuture.supplyAsync(() -> this.doGet(1), executorService);
+    }
+
+    private T doGet(int attempt) {
+        try {
+            return retriedSupplier.get();
+        } catch (final RuntimeException e) {
+            // log error, but try again (don't end loop)
+            LOGGER.error("Client <{}>: Failed to <{}>: {}.", sessionId, nameOfAction, e.getMessage());
+            waitFor(getTimeToWaitForAttempt(attempt));
+            return doGet(attempt + 1);
+        }
```

Comment:
It is better to convert this recursion into a forever-loop so that repeated retries do not cause StackOverflowError, because JVM doesn't support tail call optimization yet [(proto, 80%)](https://openjdk.java.net/projects/mlvm/subprojects.html)

Reply:
Provided updated way of handling the loop which should solve both of your change requests.